### PR TITLE
Add JSON Support on all fields

### PR DESCRIPTION
**Pull request by parsnips**
Toward resolving #423 

I have proof this works in this PR... but given my lack of familiarity with the lib, @achille-roussel I suspect you can point out where the dragons lay.

### CODE REVIEW

File: type.go [hunk 1]
```diff
@@ -1316,7 +1317,20 @@ func (t *jsonType) Decode(dst encoding.Values, src []byte, enc encoding.Encoding
 }
 
 func (t *jsonType) AssignValue(dst reflect.Value, src Value) error {
-	return ByteArrayType.AssignValue(dst, src)
+	switch dst.Kind() {
+	// Assign value using ByteArrayType for BC...
+	case reflect.String, reflect.Slice:
```
Comment (parsnips) [inline type.go:1322]: It feels like we shouldn't need to dispatch to the `ByteArrayType` at all... but i kept this in - in case I'm wrong.

**Comment by parsnips** [comment]
@achille-roussel I think this is ready for review.

**Comment by ryandeivert** [comment]
@parsnips / @achille-roussel any updates on this PR? I'm interested in this functionality, but not entire sure the current work solves for my use-case. 

I have a `map[string]interface{}` that I'd like to apply this to, and given the way #423 is outlined, I think this should be possible.

I currently have a "sanitization" step that converts `map[string]interface{}` to `map[string]string`, where the sanitized string value is the json representation of the interface (for non-string values).

I have to hard-code this sanitization for all map types now, which is very inflexible. Essentially I would like to do away with this step and have the serializer do it for me.

I have my project pointing to @parsnips feature branch (`parsnips:parsnips/full-json`), but can't seem to get a basic example using `structpb.Value` working.

Any thoughts?

Basic example of what I currently do:

```golang
type Example struct {
	FooData   map[string]interface{} `json:"fooData" parquet:"-"`
	PqFooData map[string]string      `parquet:"foodata,optional"`
	BarData   map[string]interface{} `json:"barData" parquet:"-"`
	PqBarData map[string]string      `parquet:"bardata,optional"`
}

func (c *Example) sanitize() {
	jsonCast(c.FooData, &c.PqFooData)
	jsonCast(c.BarData, &c.PqBarData)
}

func jsonCast(source map[string]interface{}, dest *map[string]string) {
	*dest = make(map[string]string)
	for k, v := range source {
		var cast string
		switch i := v.(type) {
		case string:
			cast = i // already string, do not marshal
		default:
			b, err := json.Marshal(i)
			if err != nil {
				log.Printf("could not convert to json: %s (source: %s)", err, source)
				continue
			}
			cast = string(b)
		}
		(*dest)[k] = cast
	}
}

// fileToParquet opens a file for reading and writes Example structs to parquet
func fileToParquet(file string, writer *parquet.GenericWriter[Example]) {
	in, err := os.Open(file)
	if err != nil {
		log.Fatal(err)
	}
	defer in.Close()

        // readFile returns a channel that yields Example structs
	for ex := range readFile(in) {
		ex.sanitize() // prefer to not do this
		_, err = writer.Write([]Example{ex})
		if err != nil {
			log.Printf("failed to write object to parquet: %s", err)
		}
	}
}
```

**Comment by ryandeivert** [comment]
update: I was able to solve for the above with a custom type + json marshaler at load time, but would prefer to have a more extensible way to do this if you have any suggestions. thanks for providing a great package 🥇

**Comment by parsnips** [comment]
@ryandeivert I have a repro of this - funnily enough the entire point of writing this was to encode `structpb.Value` as json 😂  I should have a fix today.

**Comment by parsnips** [comment]
@ryandeivert This should work for you out of the box now.

**Comment by parsnips** [comment]
@bartleyg can you kick off the tests again?

**Comment by ryandeivert** [comment]
thanks @parsnips! I may wait for it to merge here to test out, since I've got something _functional_ for the time being. thanks for the great addition!

**Comment by ryandeivert** [comment]
trying this out now @parsnips  -- can you provide an example of what is working for you?

I still receive the following:
```
panic: cannot create parquet node from go value of type structpb.isValue_Kind

goroutine 1 [running]:
github.com/segmentio/parquet-go.nodeOf({0x101062358?, 0x100fda940}, {0x1400007dcb0, 0x3, 0x8?})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:576 +0xd8c
github.com/segmentio/parquet-go.makeNodeOf({0x101062358, 0x100fda940}, {0x100f9a60f?, 0xd?}, {0x1400007dcb0, 0x3, 0x3})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:868 +0x384
github.com/segmentio/parquet-go.structNodeOf({0x101062358?, 0x10100f9e0})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:328 +0x268
github.com/segmentio/parquet-go.nodeOf({0x101062358?, 0x10100f9e0}, {0x1400007dc50, 0x3, 0x8?})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:572 +0xb1c
github.com/segmentio/parquet-go.makeNodeOf({0x101062358, 0x10100f9e0}, {0x100fb0b63?, 0xd?}, {0x1400007dc50, 0x3, 0x3})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:868 +0x384
github.com/segmentio/parquet-go.structNodeOf({0x101062358?, 0x100ff05a0})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:328 +0x268
github.com/segmentio/parquet-go.nodeOf({0x101062358?, 0x100ff05a0}, {0x1400007dbc0, 0x3, 0x8?})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:572 +0xb1c
github.com/segmentio/parquet-go.makeNodeOf({0x101062358, 0x100ff05a0}, {0x100fca985?, 0xd?}, {0x1400007dbc0, 0x3, 0x3})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:868 +0x384
github.com/segmentio/parquet-go.structNodeOf({0x101062358?, 0x101018be0})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:328 +0x268
github.com/segmentio/parquet-go.nodeOf({0x101062358?, 0x101018be0}, {0x1400007db60, 0x3, 0x8?})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:572 +0xb1c
github.com/segmentio/parquet-go.makeNodeOf({0x101062358, 0x101018be0}, {0x100fb7ec1?, 0xd?}, {0x1400007db60, 0x3, 0x3})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:868 +0x384
github.com/segmentio/parquet-go.structNodeOf({0x101062358?, 0x10102bfe0})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:328 +0x268
github.com/segmentio/parquet-go.nodeOf({0x101062358?, 0x10102bfe0}, {0x1400007d980, 0x3, 0x8?})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:572 +0xb1c
github.com/segmentio/parquet-go.makeNodeOf({0x101062358, 0x10102bfe0}, {0x100fb05a2?, 0xd?}, {0x1400007d980, 0x3, 0x3})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:868 +0x384
github.com/segmentio/parquet-go.structNodeOf({0x101062358?, 0x1010490e0})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:328 +0x268
github.com/segmentio/parquet-go.nodeOf({0x101062358?, 0x1010490e0}, {0x0, 0x0, 0x1400020f9f8?})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:572 +0xb1c
github.com/segmentio/parquet-go.schemaOf({0x101062358?, 0x1010490e0})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/schema.go:113 +0xbc
github.com/segmentio/parquet-go.NewGenericWriter[...]({0x1010561c0, 0x1400000e068?}, {0x1400020ff00, 0x7?, 0x100ed3f72?})
	$GOPATH/pkg/mod/github.com/segmentio/parquet-go@v0.0.0-20230330194643-f785677b9a75/writer_go18.go:84 +0xd0
main.main()
	$GOPATH/src/.../.../repo/main.go:70 +0x48c
exit status 2
```

**Comment by parsnips** [comment]
@ryandeivert the unit test in this pr demonstrates https://github.com/parsnips/parquet-go/blob/dfb4fd02d436f49670f3eda96c9d63360dea44d6/parquet_go18_test.go#L345

**Pull request merged**